### ISSUE TITLE
Use Guice Stages

### ISSUE THREAD

**Issue by guillaumecle** [cbeust/testng#663]
Guice let you chose between three stages when creating an injector: PRODUCTION, DEVELOPMENT and TOOL.
I usually run all my tests with Stage.DEVELOPMENT, mainly because Singleton are lazily created and this speed up my tests.
Could testng creates injectors with Stage.DEVELOPMENT or let it configurable through @Guice annotation?

**Comment by juherr** [comment]
As I see, `Stage` is only available for the first injector ("parent injector" in TestNG).
`@Guice` creates a new child `Injector` from the "parent".
So it will be difficult to provide configurable thing.

Maybe TestNG should use `DEVELOPMENT`. @cbeust ?

But you can try to create your own "parent injector" with the expected `Stage` too, and change it in a `ISuiteListener` via `ISuite#setParentInjector`.

**Comment by guillaumecle** [comment]
I'm trying the `ISuite#setParentInjector` way (not easy with the out of date eclipse plugin).
It seems that the "original" injector is created before `ISuiteListener#onStart` is called, isn't it?

**Comment by juherr** [comment]
You can do yourself what TestNG do with something like:

```
  @Override
  public void onStart(ISuite suite) {
    Injector injector;
    if (!suite.getParentModule().isEmpty()) {
      Class<Module> parentModule = (Class<Module>) ClassHelper.forName(suite.getParentModule());
      if (parentModule == null) {
        throw new TestNGException("Cannot load parent Guice module class: " + parentModule);
      }
      parentModule = parentModule.newInstance();
      injector = com.google.inject.Guice.createInjector(DEVELOPMENT, parentModule);
    } else {
      injector = com.google.inject.Guice.createInjector(DEVELOPMENT);
    }
    suite.setParentInjector(injector);
  }
```

For your eclipse issue, see cbeust/testng-eclipse#122, cbeust/testng-eclipse#123 and cbeust/testng-eclipse#124

**Comment by guillaumecle** [comment]
Oh yes, the `Use project TestNG jar` is really useful, even if I don't get why it's peeking the `6.8.9beta` when my project is using `6.8.21`.

This is almost exactly the `ISuiteListener#onStart` I wrote. My point is that the injector is created before `ISuiteListener#onStart` is called (and so `ISuite#getParentInjector` is called before I do my custom  `ISuite#setParentInjector`)

**Comment by juherr** [comment]
In fact, you should avoid @TestNG from your tests and doing the injection by hand:

```
  @Override
  public void onStart(ITestContext context) {
    Injector injector = context.getSuite().getParentInjector();
    for (ITestNGMethod method : context.getAllTestMethods()) {
      injector.injectMembers(method.getInstance());      
    }
  }
```

And Eclipse is not really using 6.8.9beta. See #616 ;)

**Comment by guillaumecle** [comment]
You mean avoid `@Guice`? Yes I could do that, but it would be less easy to passpass my modules or moduleFactory to my listener. I could perhaps create my custom annotation to specify them ...

But I'd loved to use `@Guice`. It seems to me like redoing some code that already exits inside TestNG and I could lose some feature like the new injection in `DataProvider`

**Comment by guillaumecle** [comment]
Could we fine another way to configure the `Stage`? Perhaps at TestSuite level?

**Comment by juherr** [comment]
Can you try to make a PR with configurable `Stage`?

**Comment by guillaumecle** [comment]
Doesn't it make sense to have some guice related config at TestSuite level? A suite has nothing to do with guice or even injection?

I got another idea. What about build the parent injector with the first stage found in the suite and if a different stage is found on another test of the suite fail it? This means force all injected tests of a suite to have the same stage.
I'll try to understand more the relation between parent and child injector and see what is possible to do.

**Comment by guillaumecle** [comment]
I was wrong, there already is some guice config at suite level ```<suite parent-module="com.example.SuiteParenModule">``` I'll add the ```Stage``` there, something like ```stage="myStage"```

And what should be the default stage? I vote for ```DEVELOPMENT```

**Comment by juherr** [comment]
As stage is a common word, maybe something like `guice-stage` is better. 

About the default stage, maybe @cbeust can give us his opinion?

**Comment by cbeust** [comment]
Yes, `guice-stage` is probably better. I wish I had prefixed the module similarly (`guice-parent-module`) for consistency.

**Comment by juherr** [comment]
But you prefer PRODUCTION or DEVELOPMENT as default scope? 

The current one is PRODUCTION but @guillaumecle asked to change it to DEVELOPMENT what is a good idea IMHO. But maybe you see bad side effect?

**Comment by cbeust** [comment]
No objection to DEVELOPMENT.

**Comment by guillaumecle** [comment]
I did it on [my fork](https://github.com/guillaumecle/testng) and I realized that the default ```Stage``` of Guice is ```DEVELOPMENT``` and not ```PRODUCTION``` as I thought.

Do you want me to do a pull request?

**Comment by cbeust** [comment]
Sure!

-- 
Cédric

**Comment by juherr** [comment]
Fixed by #682
